nx/lib/nx/shared.ex

```diff
@@ -387,4 +387,35 @@ defmodule Nx.Shared do
             "#{inspect(struct1)} and #{inspect(struct2)}. You may need to call Nx.backend_transfer/1 " <>
             "(or Nx.backend_copy/1) on one or both of them to transfer them to a common implementation"
   end
+
+  @doc """
+  Used to define an Nx callback with an optional implementation.
+
+  The given body is used as the default implementation otherwise.
+  """
+  defmacro defoptional(function_header, body) do
```

Comment:
@josevalim I ended up trying to implement a macro for this, but I think that perhaps this isn't the way.

For one, I couldn't figure out how to call the `_implementation` defp from within the macro.
Secondly, I'm a bit confused on the expected arity for the optional callback. Normally when we define callbacks there's also the expected output holder.

Should we force the functions to define their expected outputs so it can be passed as  `[expected_output_holder | args]` to the `apply/3` call below?

Other than this, the code is currently working as expected, using the Torchx pseudo implementation when the given tensors are from Torchx.Backend and calling the default one for Nx.BinaryBackend

Reply:
I'm thinking about reverting to the original suggestion like below, but with the addition of the output holder:

```elixir
def optional(function_name, args, default_impl) do
  arity = length(args)
  backend = impl!(args)
  cond do
    backend == Nx.Defn.Expr -> raise "not implemented"
    function_exported?(backend, name, arity + 1) -> 
      apply(backend, name, args)
    :otherwise -> 
      apply(default_impl, args)
  end
end
```

Which can be used like so:

```elixir
def solve(a, b) do
  # keep the input validations here
  output_holder = b
  optional(:solve, [output_holder, a, b], fn out, a, b -> 
    # main linalg code here
  end)
end
```

Reply:
Yeah, I wouldn't start with the macro, it feels backward. Let's first start with the building blocks and then add macro functionality on top if required.

Reply:
I think the output holder be part of the args, yeah, because that's what we would call in the backend.

Reply:
Actually, I am not sure about the output. We have three options:

1. say the output is not included for optional callbacks
2. automatically compute the output by applying the anonymous functions with a tensor expression
3. require the output explicitly (which most likely requires doing 2 anyway)